dilation-protocol.md

```diff
@@ -1,371 +1,280 @@
+
+All subsequent messages as known as "Noise transport messages", and are independent for each direction.
+Transport messages are encrypted by the shared key, in a form that evolves as more messages are sent.
+
+The Follower's first transport message is an empty packet, which we use as a "key confirmation message" (KCM).
+
+The Leader doesn't send a transport message right away: it waits to see the Follower's KCM,
+which indicates this connection is viable (i.e. the Follower used the same dilation key as the Leader,
+which means they both used the same wormhole code).
+
+Multiple connection attempts and handshakes may happen in parallel.
+Of those where the Leader received a KCM, it chooses one of them to be actually used.
+It will send an empty KCM to the Follower to mark that connection, and close all others.
+All other connection attempts on both sides will be cancelled.
+All listening sockets may or may not be shut down (TODO: think about it ~warner).
+
+### Messages over the Dilation connection
+
+Unlike with applications using Transit,
+the application layer messages are not sent directly over the Dilation connection.
+Instead, they get wrapped by this middle layer which provides the following features:
+
+- Sub-channel multiplexing
+- Pings and ACKs for connection status tracking
+- Automatic reconnection after connection loss
+- Durable channel with automatic re-transmission of lost messages after reconnect
+
+~~In the future, we might have L2 links that are less connection-oriented,
+which might have a unidirectional failure mode, at which point we'll need to
+monitor full roundtrips. To accomplish this, the Leader will send periodic
+unconditional PINGs, and the Follower will respond with PONGs. If the
+Leader->Follower connection is down, the PINGs won't arrive and no PONGs will
+be produced. If the Follower->Leader direction has failed, the PONGs won't
+arrive. The delivery of both will be delayed by actual data, so the timeouts
+should be adjusted if we see regular data arriving.~~
+
+If the Leader loses the Dilation connection for whatever reason, it will initiate a reconnection attempt.
+This involves sending a `reconnect` message and according hints over the Mailbox connection,
+the same way as for initially establishing the connection.
+Once a new connection is established and the handshake completes,
+both sides re-transmit their outbound messages that were lost due to the connection interruption.
+
+Each message over the Dilation connection starts with a one-byte type indicator.
+The rest of the message depends upon the type:
+
+* 0x00 PING, 4-byte ping-id
+* 0x01 PONG, 4-byte ping-id
+* 0x02 OPEN, 4-byte subchannel-id, 4-byte seqnum
+* 0x03 DATA, 4-byte subchannel-id, 4-byte seqnum, variable-length payload
+* 0x04 CLOSE, 4-byte subchannel-id, 4-byte seqnum
+* 0x05 ACK, 4-byte response-seqnum
+
+TODO it would make more sense to have the seqnum before the subchannel-id, processing wise ~piegames
```

Comment:
Why? Also you have to deserialize all of this to process it, so I can't see why order matters at all.

Reply:
I agree that it mostly doesn't matter, this is just me being pedantic.